vhost-user-backend/src/bitmap.rs

```diff
@@ -0,0 +1,615 @@
+// Copyright (C) 2024 Red Hat, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::ops::Index;
+use std::os::fd::{AsRawFd, BorrowedFd};
+use std::sync::atomic::{AtomicU8, Ordering};
+use std::sync::{Arc, RwLock};
+use std::{io, ptr};
+use vm_memory::bitmap::{Bitmap, BitmapSlice, WithBitmapSlice};
+use vm_memory::mmap::NewBitmap;
+use vm_memory::{Address, GuestMemoryRegion};
+
+// Size in bytes of the `VHOST_LOG_PAGE`
+const LOG_PAGE_SIZE: usize = 0x1000;
+// Number of bits grouped together as a basic storage unit ("word") in the bitmap.
+const LOG_WORD_SIZE: usize = u8::BITS as usize;
+
+/// A `Bitmap` with an internal `Bitmap` that can be replaced at runtime
+pub trait BitmapReplace: Bitmap {
+    type InnerBitmap: MemRegionBitmap;
+
+    /// Replace the internal `Bitmap`
+    fn replace(&self, bitmap: Self::InnerBitmap);
+}
+
+/// A bitmap relative to a memory region
+pub trait MemRegionBitmap: Sized {
+    /// Creates a new bitmap relative to `region`, using the `logmem` as
+    /// backing memory for the bitmap
+    fn new<R: GuestMemoryRegion>(region: &R, logmem: Arc<MmapLogReg>) -> io::Result<Self>;
```

Comment:
nit: If this trait only serves as a constructor then I find the name a bit generic. `MemRegionBitmap` sounds like something that I can do something with, rather than something that I merely use as contract to construct something else?

But also see my other comment on why we need these traits at all.

Reply:
I'll respond in the last comment